src/plugins/data/common/index_patterns/fields/field_list.ts

```diff
@@ -18,95 +18,110 @@
  */
 
 import { findIndex } from 'lodash';
-import { IIndexPattern } from '../../types';
-import { IFieldType } from '../../../common';
-import { Field } from './field';
+import { IFieldType, shortenDottedString } from '../../../common';
+import { IndexPatternField } from './index_pattern_field';
 import { OnNotification, FieldSpec } from '../types';
-import { FieldFormatsStartCommon } from '../../field_formats';
+import { IndexPattern } from '../index_patterns';
 
-type FieldMap = Map<Field['name'], Field>;
+type FieldMap = Map<IndexPatternField['name'], IndexPatternField>;
 
-interface FieldListDependencies {
-  fieldFormats: FieldFormatsStartCommon;
-  onNotification: OnNotification;
-}
-
-export interface IIndexPatternFieldList extends Array<Field> {
-  getByName(name: Field['name']): Field | undefined;
-  getByType(type: Field['type']): Field[];
+export interface IIndexPatternFieldList extends Array<IndexPatternField> {
   add(field: FieldSpec): void;
+  getByName(name: IndexPatternField['name']): IndexPatternField | undefined;
+  getByType(type: IndexPatternField['type']): IndexPatternField[];
   remove(field: IFieldType): void;
+  removeAll(): void;
+  replaceAll(specs: FieldSpec[]): void;
   update(field: FieldSpec): void;
 }
 
 export type CreateIndexPatternFieldList = (
-  indexPattern: IIndexPattern,
+  indexPattern: IndexPattern,
   specs?: FieldSpec[],
-  shortDotsEnable?: boolean
+  shortDotsEnable?: boolean,
+  onNotification?: OnNotification
 ) => IIndexPatternFieldList;
 
-export const getIndexPatternFieldListCreator = ({
-  fieldFormats,
-  onNotification,
-}: FieldListDependencies): CreateIndexPatternFieldList => (...fieldListParams) => {
-  class FieldList extends Array<Field> implements IIndexPatternFieldList {
-    private byName: FieldMap = new Map();
-    private groups: Map<Field['type'], FieldMap> = new Map();
-    private indexPattern: IIndexPattern;
-    private shortDotsEnable: boolean;
-    private setByName = (field: Field) => this.byName.set(field.name, field);
-    private setByGroup = (field: Field) => {
-      if (typeof this.groups.get(field.type) === 'undefined') {
-        this.groups.set(field.type, new Map());
-      }
-      this.groups.get(field.type)!.set(field.name, field);
-    };
-    private removeByGroup = (field: IFieldType) => this.groups.get(field.type)!.delete(field.name);
+export class FieldList extends Array<IndexPatternField> implements IIndexPatternFieldList {
```

Comment:
I'm a bit concerned about extending the native `Array` here. `Array` has many *public* methods, are we sure we want to expose all of those as part of `FieldList`. If users use those, they can basically mutate the `FieldList` while by-passing the business logic of `FieldList`.

Maybe we could use composition instead?

```ts
class FieldList {
  private data: Array<IndexPatternField> = [];
}
```

Reply:
++

Reply:
100% agree, but I'm avoiding changing the class interfaces in this PR. Will address in a follow up.